Respect more configuration options

laughedelic opened the pull request:
This fixes some points in #161:
- respect hover.enabled for tooltips on hover
- respect scalac.enabled for diagnostics, completions and signature help
- add an option to control symbol highlights (`textDocument.documentHighlight`)

Some things for discussion:
- I think that the features provided by presentation compiler should have their own switches. Although their stability depends on the state of PC, even if they do work, they may work differently and have different value for the user. So user may want to enable only some of them. 
- I got confused by the `PublishSquigglies` vs. `PublishScalacDiagnostics`. Isn't it all just diagnostics? Also I'm not sure I handled `scalac.enabled` for the diagnostics well. If I disable `sbt.enabled` and `scalac.enabled`, but enable `scalafix.enabled`, I still get "squigglies" on save — where do they come from? Does it mean that PC still generates semanticdb for scalafix?
- I disabled symbol highlights by default because it works well in read-only mode, but breaks immediately once I start changing code. This is something to be diagnosed and addressed separately, but I think this feature is not stable enough to be enabled by default.

olafurpg commented on metals/src/main/scala/scala/meta/metals/ScalametaServices.scala:
```diff
@@ -305,11 +313,13 @@ class MetalsServices(
-        Uri(params.textDocument.uri),
-        params.position
-      )
+      if (latestConfig().highlight.enabled) {
```
Have you tried to dynamically unregister capabilities? I think it makes sense to keep this guard regardless because I'm not sure if we can rely on it in case a client like vim/sublime to respect dynamic capabilities. Just curious to see if it works.

laughedelic replied:
Sorry, didn't get what you're saying here. Do you really mean dynamically unregister capabilities or you mean just enable/disable this setting in the config?

olafurpg replied:
I was thinking of using `client/unregisterCapability` https://microsoft.github.io/language-server-protocol/specification#client_unregisterCapability

I think it's best to do it like here in the PR because dynamic capabilities may not be supported by all clients, but I was just wondering if you had tried to used `unregisterCapabilities`

laughedelic replied:
Ah, now I understand: when some feature gets enabled/disabled in the config, we should register/unregister the corresponding capability. 

I agree that it's correct, but we don't register capabilities dynamically yet and you're right that it's not widely supported (Atom doesn't support it, for instance), so I'd suggest we address it later, probably after #163.

olafurpg replied:
I agree this should be explored later, esp. if it's not even supported by atom 👍

olafurpg commented on metals/src/main/scala/scala/meta/metals/Configuration.scala:
```diff
@@ -30,6 +31,7 @@ object Configuration {
   )
   @JsonCodec case class Scalac(enabled: Boolean = false)
   @JsonCodec case class Hover(enabled: Boolean = false)
```
Unrelated to this PR. I think it might be safe to enable hover by default. It should work as robustly if more robustly than goto definition

olafurpg commented:
> I got confused by the PublishSquigglies vs. PublishScalacDiagnostics. Isn't it all just diagnostics? 

Good question, I think this is a relic from when we separated scalac and scalafix diagnostics causing them to race with each other. We should probably just have a single `PublishDiagnostics` effect.

olafurpg commented:
> I think that the features provided by presentation compiler should have their own switches. Although their stability depends on the state of PC, even if they do work, they may work differently and have different value for the user. So user may want to enable only some of them.

I agree! How about `scalac.diagnostics = true/false; scalac.completions = true/false`?

laughedelic commented:
@olafurpg Thanks for the review! I addressed your comments:
* `scalac.completions.enabled` and `scalac.diagnostics.enabled`
* if no PC features enabled, no interactive semanticdbs
* `PublishSquigglies` & `PulblishScalacDiagnostics` –> `PulblishDiagnostics`

Let's deal with the hover/highlights improvements in another PR.

olafurpg commented on metals/src/main/scala/scala/meta/metals/Configuration.scala:
```diff
@@ -28,8 +29,15 @@ object Configuration {
       command: String = "test:compile"
   )
-  @JsonCodec case class Scalac(enabled: Boolean = false)
+  @JsonCodec case class Scalac(
```
Might be nice to add a `def enabled: Boolean = completions.enabled || diagnostics.enabled`

laughedelic replied:
Added in 15398a1

olafurpg commented on metals/src/main/scala/scala/meta/metals/ScalametaServices.scala:
```diff
@@ -87,13 +89,16 @@ class MetalsServices(
-          .fromIterable(Semanticdbs.toSemanticdb(input, scalacProvider))
-          .executeOn(presentationCompilerScheduler)
-      }
+    if (latestConfig().scalac.completions.enabled ||
```
I would move this condition into the `flatMap` so that we can respond to dynamic updates on the configuration. I think otherwise `val interactiveSemanticdbs` will always be `Observable.empty`

laughedelic replied:
Done in 15398a1

laughedelic commented:
> Unfortunately we don't have many integration tests for configuration, any ideas how we could improve that?

To be honest, I didn't think about integration tests for the config yet. Do you have any ideas?
